### PR TITLE
Allow using PPXes as libraries 

### LINKED ISSUE #121: RFC: support to ppxlib directly on bsconfig
Currently ppx need to be built into a standalone binary to work with melange, but as dune we could add support to ppxlib for now.

## Pros

- Faster when multiple ppx is present because it is a single pass
- No need to additional binaries in the env

## Cons

- It's a bsb feature which will hopefully not be long lived

### PR DESCRIPTION
**Pull request by anmonteiro**
fixes #121 

- generates a ppxlib driver on the fly

#### TODO:

- [x] allow passing arguments to the ppx

### CODE REVIEW

File: jscomp/bsb/bsb_merlin_gen.ml [hunk 1]
```diff
@@ -141,7 +141,11 @@ let merlin_file_gen ~per_proj_dir:(per_proj_dir:string)
   if generate_merlin then begin
     let buffer = Buffer.create 1024 in
     output_merlin_namespace buffer namespace;
-    Ext_list.iter ppx_files (fun ppx ->
+    if ppx_config.ppxlib <> [] then begin
+      Buffer.add_string buffer merlin_flg_ppx;
+      Buffer.add_string buffer ".melange.eobjs/ppx.exe"
```
Comment (EduardoRFS) [inline jscomp/bsb/bsb_merlin_gen.ml:146]: **warning** hard coded string

File: jscomp/bsb/bsb_ninja_gen.ml [hunk 2]
```diff
@@ -169,7 +175,13 @@ let output_ninja_and_namespace_map
   match package_kind with
   | Bsb_package_kind.Toplevel ->
     (* emitted in bsb_main *)
```
Comment (EduardoRFS) [inline jscomp/bsb/bsb_ninja_gen.ml:177]: So this comment is now invalid?
```suggestion
```

**Review by EduardoRFS — CHANGES_REQUESTED** [review]
Two small nits

**Comment by TheSpyder** [comment]
How do I make use of this? It's not really clear from looking at the changes.

**Comment by anmonteiro** [comment]
@TheSpyder this will definitely need documented, and we've added it to the list in #105.

A few informal notes:

1. depend on the PPXes you want via esy (or OPAM or Nix, anything that knows how to install PPXes with the ocamlfind toolchain)
2. add them to `ppx-flags` as usual, with just the ppx name (this is the important part -- it should match whatever the findlib name is. Generally it's also the package name)
3. Profit!

**Comment by TheSpyder** [comment]
That's the part I'm not clear on, the findlib name. I'm trying to make a branch of my old favourite [decco](https://github.com/reasonml-labs/decco) that can integrate with this feature, to measure the speed difference on my fairly large project.

**Comment by anmonteiro** [comment]
you can run `esy ls-libs` I believe, which will tell you the name of the library.

**Comment by anmonteiro** [comment]
You need to depend on decco from source, btw, in your case via esy. I don't know how easy this is, but we could make a PR.

**Comment by TheSpyder** [comment]
yes, I tried that and it still tried to execute it as a binary instead of loading it as a library. I just need to fiddle a bit more (I'm currently trying to lift the `esy` project up to the respository root instead of having it in a subfolder, to see if that helps).

**Comment by TheSpyder** [comment]
I can't get it working, [here's the branch I tried](https://github.com/thespyder/decco/tree/melange), but I also realised that this "PPX as library" feature really only helps if there are multiple PPX. It still builds the ppx as a separate executable, so for single ppx use it won't be any faster than the precompiled binary (just easier to install since precompiled executables are not required).

This feature will likely help me in the future, but it's no longer important for me to try and get it working today :)

**Comment by EduardoRFS** [comment]
@TheSpyder yeah that's true for us, but there is always multiple ppx pass, because we do a ppx pass ourselves.

Ideally we should move Melange PPX to ppxlib

**Pull request merged**